lib/scope.js

```diff
@@ -413,6 +429,147 @@ module.exports = {
     return winner;
   },  // Ends scope.tapElement()
 
+  // Handle server reload errors
+  server_statuses: [],
+  server_check_interval_ms: 1 * 1000,
+  stop_tracking_server_status: false,
+  track_server_status: async function( scope ) {
+    /** Updates server_statuses list by checking if server is responding until,
+    *    eventually, it is told to stop. */
+
+    while ( !scope.stop_tracking_server_status ) {
+      // This code assumes a reasonable response time would be 2 seconds if the
+      // server is up and that a server reload/restart won't take less than 2 seconds.
+      // Also, it's non-blocking, so it doesn't slow down tests
+      let is_up = await server.is_server_responding({ timeout: scope.server_check_interval_ms });
+      // Add status to the beginning of the list to make truncating easier
+      scope.server_statuses.unshift( is_up );
+
+      // To make it easier to check past server statuses, remove defunct server statuses
+      let max_length = Math.round( scope.timeout/scope.server_check_interval_ms );
+      // Trying to extract more items than the actual length is ok
+      scope.server_statuses = scope.server_statuses.slice( 0, max_length );
```

Comment:
Just to clarify, we want exactly 1 element per second in here, as opposed to just a constant number of elements?

Reply:
Yes. I was assuming my 'waiting' math was right for checking once per second, but as you've spotted, thankfully, that particular wait time wasn't doing what I thought it was.

This is part of the previous comment, right? - one about the `is_server_responding` timeout? If there's another issue to address here, let me know.

Reply:
It's separate from that; I'm fairly confident that we do want one status per second. This question was specifically if we wanted "wait time" number of elements in the statuses array, or if we just wanted a fixed number of things in there. If we always have i.e. 30 elements in the array, no matter what `scope.timeout` is, then the code gets a bit simpler, that's all.

Reply:
The number of elements does need to cover the whole timeout that the developer wants. If they gave a timeout of 2 min, I believe we need to check whether the server reloaded during any of that time.

I'm not sure I can find a good way to make an example, but I'll try. Suppose we do only store the last 30 items in a hard-coded way. Suppose the reload happened during the first 15 seconds of it, leaving the "user" hanging out waiting for a timeout screen, and then the server managed to finish reloading during the next minute, and it then left the "user" with an error screen later or some such (one of our servers waits 5 min or something). If we only store the last 30 seconds, it would look like the server was up that whole time and that reloading wasn't part of the issue. If we store the whole 2 min, we'd see that the server did in fact reload and this error could be a reload issue.

Not sure if that made any sense.

Reply:
I get the idea, thanks. But this part made me more confused about other things:

> Suppose the reload happened during the first 15 seconds of it, leaving the "user" hanging out waiting for a timeout screen, and then the server managed to finish reloading during the next minute, and it then left the "user" with an error screen later or some such (one of our servers waits 5 min or something)

Is this saying that if there's some / any error we'll just restart, assuming it was caused by the reload? So if there is some non-deterministic error happening but it occurs close to a restart it'll be lost?

Reply:
No, sorry that's not clear. This all only pops off if there's a timeout error.  If a server reload was detected during the timeout, the developer will be informed that the server was reloading. Otherwise, they will get a different timeout error message. I believe that for a timeout error, if the server wasn't reloading there is nothing else we can detect. There's always a screenshot if no secret values were used. Can you think of more we can do or detect for timeout errors?

Reply:
Gotcha, apologies for the confusion. I can't think of any additional information that we could get there.